src/main/java/com/minecolonies/coremod/client/gui/WindowScan.java

```diff
@@ -0,0 +1,318 @@
+    }
+
+    /**
+     * Clears and resets/updates all resources.
+     */
+    private void updateResources()
+    {
+        try
+        {
+            final int x1 = Integer.parseInt(pos1x.getText());
+            final int y1 = Integer.parseInt(pos1y.getText());
+            final int z1 = Integer.parseInt(pos1z.getText());
+            pos1 = new BlockPos(x1, y1, z1);
+
+            final int x2 = Integer.parseInt(pos2x.getText());
+            final int y2 = Integer.parseInt(pos2y.getText());
+            final int z2 = Integer.parseInt(pos2z.getText());
+            pos2 = new BlockPos(x2, y2, z2);
+        }
+        catch(final NumberFormatException e)
+        {
+            Minecraft.getMinecraft().player.sendMessage(new TextComponentString("Invalid Number - Closing!"));
+            close();
+        }
+        
+        final World world = Minecraft.getMinecraft().world;
+        resources.clear();
+
+        for(int x = Math.min(pos1.getX(), pos2.getX()); x <= Math.max(pos1.getX(), pos2.getX()); x++)
+        {
+            for(int y = Math.min(pos1.getY(), pos2.getY()); y <= Math.max(pos1.getY(), pos2.getY()); y++)
+            {
+                for(int z = Math.min(pos1.getZ(), pos2.getZ()); z <= Math.max(pos1.getZ(), pos2.getZ()); z++)
+                {
+                    final BlockPos here = new BlockPos(x, y, z);
+                    final IBlockState blockState = world.getBlockState(here);
+                    final TileEntity tileEntity = world.getTileEntity(here);
+                    final List<Entity> list = world.getEntitiesWithinAABB(Entity.class, new AxisAlignedBB(here));
+
+                    for (final Entity entity : list)
+                    {
+                        for (final ItemStack stack : ItemStackUtils.getListOfStackForEntity(entity, Minecraft.getMinecraft().player))
+                        {
+                            if (!ItemStackUtils.isEmpty(stack))
+                            {
+                                addNeededResource(stack, 1);
+                            }
+                        }
+
+                        if (entity instanceof EntityCitizen)
+                        {
+                            addNeededResource(new ItemStack(Blocks.MOB_SPAWNER), 1);
+                        }
+                    }
+
+                    if (here == null)
+                    {
+                        continue;
+                    }
+
+                    @Nullable final Block block = blockState.getBlock();
+                    if (block != null && block != Blocks.AIR)
+                    {
+                        if (tileEntity != null)
+                        {
+                            final List<ItemStack> itemList = new ArrayList<>();
+                            itemList.addAll(ItemStackUtils.getItemStacksOfTileEntity(tileEntity.writeToNBT(new NBTTagCompound()), world));
+                            for (final ItemStack stack : itemList)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 5 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)